src/as-suggested.c

```diff
@@ -0,0 +1,214 @@
+	if (kind == AS_SUGGESTED_KIND_UPSTREAM)
+		return "upstream";
+
+	return "unknown";
+}
+
+/**
+ * as_suggested_kind_from_string:
+ * @kind_str: the string.
+ *
+ * Converts the text representation to an enumerated value.
+ *
+ * Returns: a #AsSuggestedKind or %AS_SUGGESTED_KIND_UNKNOWN for unknown
+ **/
+AsSuggestedKind
+as_suggested_kind_from_string (const gchar *kind_str)
+{
+	if (g_strcmp0 (kind_str, "heuristic") == 0)
+		return AS_SUGGESTED_KIND_HEURISTIC;
+	if (g_strcmp0 (kind_str, "upstream") == 0)
+		return AS_SUGGESTED_KIND_UPSTREAM;
+
+	return AS_SUGGESTED_KIND_UNKNOWN;
+}
+
+/**
+ * as_suggested_finalize:
+ **/
+static void
+as_suggested_finalize (GObject *object)
+{
+	AsSuggested *suggested = AS_SUGGESTED (object);
+	AsSuggestedPrivate *priv = GET_PRIVATE (suggested);
+
+	g_ptr_array_unref (priv->cpts_id);
+
+	G_OBJECT_CLASS (as_suggested_parent_class)->finalize (object);
+}
+
+/**
+ * as_suggested_init:
+ **/
+static void
+as_suggested_init (AsSuggested *suggested)
+{
+	AsSuggestedPrivate *priv = GET_PRIVATE (suggested);
+
+	priv->cpts_id = g_ptr_array_new_with_free_func ((GDestroyNotify) g_object_unref);
+}
+
+/**
+ * as_suggested_new:
+ *
+ * Creates a new #AsSuggested.
+ *
+ * Returns: (transfer full): a new #AsSuggested
+ **/
+AsSuggested*
+as_suggested_new (void)
+{
+	AsSuggested *suggested;
+	suggested = g_object_new (AS_TYPE_SUGGESTED, NULL);
+	return AS_SUGGESTED (suggested);
+}
+
+/**
+ * as_suggested_class_init:
+ **/
+static void
+as_suggested_class_init (AsSuggestedClass *klass)
+{
+	GObjectClass *object_class = G_OBJECT_CLASS (klass);
+	object_class->finalize = as_suggested_finalize;
+}
+
+/**
+ * as_suggested_get_kind:
+ * @suggested: a #AsSuggested instance.
+ *
+ * Gets the suggested kind.
+ *
+ * Returns: the #AssuggestedKind
+ **/
+AsSuggestedKind
+as_suggested_get_kind (AsSuggested *suggested)
+{
+	AsSuggestedPrivate *priv = GET_PRIVATE (suggested);
+	return priv->kind;
+}
+
+/**
+ * as_suggested_set_kind:
+ * @suggested: a #AsSuggested instance.
+ * @kind: the #AsSuggestedKind, e.g. %AS_SUGGESTED_KIND_HEURISTIC.
+ *
+ * Sets the suggested kind.
+ **/
+void
+as_suggested_set_kind (AsSuggested *suggested, AsSuggestedKind kind)
+{
+	AsSuggestedPrivate *priv = GET_PRIVATE (suggested);
+	priv->kind = kind;
+}
+
+/**
+ * as_suggested_get_components_id:
+ * @suggested: a #AsSuggested instance.
+ *
+ * Get a list of components id that generated the suggestion
+ *
+ * Returns: an array of components id
+ */
+GPtrArray*
+as_suggested_get_components_id (AsSuggested* suggested)
+{
+	AsSuggestedPrivate *priv = GET_PRIVATE (suggested);
+
+	return priv->cpts_id;
+}
+
+
+/**
+ * as_suggested_add_component_id:
+ * @suggested: a #AsSuggested instance.
+ * @cpt_id: The component id to add
+ *
+ * Add a component id to this suggested object.
+ **/
+void
+as_suggested_add_component_id (AsSuggested *suggested, gchar* cpt_id)
+{
+	AsSuggestedPrivate *priv = GET_PRIVATE (suggested);
+	g_ptr_array_add (priv->cpts_id, cpt_id);
```

Comment:
Uuuh, you need to strdup here, otherwise ownership of the pointer is transferred to the ptr-array, which will free it later. This isn't good.